examples/main/main.cpp

```diff
@@ -344,13 +351,28 @@ int main(int argc, char ** argv) {
         // check if we should prompt the user for more
         if (params.interactive && (int) embd_inp.size() <= n_consumed) {
 
-            // check for reverse prompt
-            if (params.antiprompt.size()) {
-                std::string last_output;
+            std::string last_output;
+            if (params.antiprompt.size() || params.stop_keywords.size()) {
                 for (auto id : last_n_tokens) {
                     last_output += llama_token_to_str(ctx, id);
                 }
+            }
+
+            // Check for stop keywords, a configurable alternative to the end-of-text token
```

Comment:
we should check the token id instead of the string for stop.

Reply:
For context, the precedent set by https://github.com/ggerganov/llama.cpp/pull/330 is to check for the string (in reverse prompts). I think checking for tokens caused a bug, or at least unintuitive behavior (https://github.com/ggerganov/llama.cpp/issues/292).

Reply:
Indeed, this should work in the same way as "antiprompts" to provide a better UX to users. Users should be able to add to the CLI parameters `--stop "### Assistant"` (for example, in the spirit of the trained `vicuna` model) or `--stop PAUSE` (for example, to implement Simon Willinson's [ReAct Python example](https://til.simonwillison.net/llms/python-react-pattern)), even though these are multi-token markers.

Reply:
Yes, see https://github.com/ggerganov/llama.cpp/issues/292#issuecomment-1476318351

Reply:
I also noticed that the stop words weren't always consistently formatted. Sometimes it would do `STOP` instead of `stop`. I'd appreciate a normalize function to down case before comparing.